Lib/test/test_httpservers.py

```diff
@@ -418,6 +418,28 @@ def test_undecodable_filename(self):
         self.check_status_and_reason(response, HTTPStatus.OK,
                                      data=os_helper.TESTFN_UNDECODABLE)
 
+    def test_get_dir_redirect_location_domain_injection_bug(self):
+        """Ensure //evil.co/..%2f../../X does not put //evil.co/ in Location.
+
+        //domain/ in a Location header is a redirect to a new domain name.
+        https://github.com/python/cpython/issues/87389
+
+        This checks that a path resolving to a directory on our server cannot
+        resolve into a redirect to another server telling it that the
+        directory in question exists on the Referrer server.
+        """
+        os.mkdir(os.path.join(self.tempdir, 'existing_directory'))
+        # Canonicalizes to /tmp/tempdir_name/existing_directory which does
+        # exist and is a dir, triggering the 301 redirect and former bug.
+        attack_url = f'//python.org/..%2f..%2f..%2f..%2f..%2f../%0a%0d/../{self.tempdir_name}/existing_directory'
```

Comment:
I suggest to build the whole expected location URL without lstrip(), for readability of the test itself:

```suggestion
        url = f'/python.org/..%2f..%2f..%2f..%2f..%2f../%0a%0d/../{self.tempdir_name}/existing_directory'
        attack_url = f'/{url}'  # //python.org/... (double slash prefix, no trailing slash)
        expected_location = f'{url}/'  # /python.org/.../ (single slash prefix, with trailing slash)
```

And replace the two asserts with just:

```
self.assertEqual(location, expected_location)
```

If you want to elaborate, you can add a comment, I don't think that it's needed to elaborate in the 3rd assertEqual() msg parameter.